src/datasets/utils/scientificMetadata.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  PartialUpdateDerivedDatasetObsoleteDto,
+  UpdateDerivedDatasetObsoleteDto,
+} from "../dto/update-derived-dataset-obsolete.dto";
+
+type DatasetDto =
+  | CreateDatasetDto
+  | CreateDerivedDatasetObsoleteDto
+  | CreateRawDatasetObsoleteDto
+  | UpdateDatasetDto
+  | UpdateDerivedDatasetObsoleteDto
+  | UpdateRawDatasetObsoleteDto
+  | PartialUpdateDatasetDto
+  | PartialUpdateDerivedDatasetObsoleteDto
+  | PartialUpdateRawDatasetObsoleteDto;
+
+type ValidatedDto = DatasetDto & { scientificMetadataValid?: boolean };
+
+@Injectable()
+export class ScientificMetadataValidator {
+  constructor(private readonly httpService: HttpService) {}
+
+  patchedMetadata(
+    datasetDto:
+      | PartialUpdateDatasetDto
+      | PartialUpdateDerivedDatasetObsoleteDto
+      | PartialUpdateRawDatasetObsoleteDto,
+    foundMetadata:
+      | CreateDatasetDto
+      | CreateDerivedDatasetObsoleteDto
+      | CreateRawDatasetObsoleteDto,
+  ):
+    | CreateDatasetDto
+    | CreateDerivedDatasetObsoleteDto
+    | CreateRawDatasetObsoleteDto {
+    const updatedDto = { ...datasetDto };
+    // Use existing metadata if not provided in the update
+    updatedDto.scientificMetadata =
+      datasetDto.scientificMetadata ?? foundMetadata.scientificMetadata;
+    updatedDto.scientificMetadataSchema =
+      datasetDto.scientificMetadataSchema ??
+      foundMetadata.scientificMetadataSchema;
+    return updatedDto as
+      | CreateDatasetDto
+      | CreateDerivedDatasetObsoleteDto
+      | CreateRawDatasetObsoleteDto;
+  }
+  async validate(
+    scientificMetadata: Record<string, unknown>,
+    scientificMetadataSchema: string,
+  ): Promise<boolean> {
+    try {
+      const response = await firstValueFrom(
+        this.httpService.get<Record<string, unknown>>(
+          scientificMetadataSchema,
+          { validateStatus: () => true },
+        ),
+      );
+
+      // Check HTTP status
+      if (response.status !== 200) {
+        Logger.log(
+          `Schema fetch failed with status ${response.status}: ${response.statusText}`,
+          "ScientificMetadataValidation",
+        );
+        return false;
+      }
+
+      const schema = response.data;
+
+      // Check if response is a valid object
+      if (!schema || typeof schema !== "object" || Array.isArray(schema)) {
+        Logger.log(
+          "Fetched schema is not a valid JSON object.",
+          "ScientificMetadataValidation",
+        );
+        return false;
+      }
+
+      const validator = new Validator();
+      const validationResult = validator.validate(scientificMetadata, schema);
+
+      return validationResult.errors.length === 0;
+    } catch (error) {
+      Logger.log(
+        error instanceof Error ? error.message : String(error),
+        "ScientificMetadataValidation",
+      );
+      return false;
+    }
+  }
+
+  async addValidationStatus(datasetDto: DatasetDto): Promise<ValidatedDto> {
+    const updatedDto = { ...datasetDto };
```

Comment:
this is a shallow copy, which I guess is fine in this case, but one should pay attention not to modify nested mutable fields. To stay on the safe side, one can use `structuredClone`